pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -90,17 +101,62 @@ func add(ctx context.Context, mgr manager.Manager, r reconcile.Reconciler) error
 		},
 	}
 
+	predTaskRun := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			o := e.ObjectOld.(*v1beta1.TaskRun)
+			n := e.ObjectNew.(*v1beta1.TaskRun)
+			// Process an update event when the old TR resource is not yet started and the new TR resource is in Pending state.
+			// This allow us to catch the PENDING condition.Reason in the BuildRun
+			if o.Status.StartTime.IsZero() && n.Status.GetCondition(apis.ConditionSucceeded).Reason == pendingReason {
+				return true
+			}
+
+			// Process an update event for every change in the condition.Reason between the old and new TR resource
+			if o.Status.GetCondition(apis.ConditionSucceeded) != nil && n.Status.GetCondition(apis.ConditionSucceeded) != nil {
+				if o.Status.GetCondition(apis.ConditionSucceeded).Reason != n.Status.GetCondition(apis.ConditionSucceeded).Reason {
+					return true
+				}
+			}
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Evaluates to false if the object has been confirmed deleted.
+			return !e.DeleteStateUnknown
+		},
+	}
+
 	// Watch for changes to primary resource BuildRun
-	err = c.Watch(&source.Kind{Type: &buildv1alpha1.BuildRun{}}, &handler.EnqueueRequestForObject{}, pred)
+	err = c.Watch(&source.Kind{Type: &buildv1alpha1.BuildRun{}}, &handler.EnqueueRequestForObject{}, predBuildRun)
 	if err != nil {
 		return err
 	}
 
-	// Watch TaskRuns
-	return c.Watch(&source.Kind{Type: &v1beta1.TaskRun{}}, &handler.EnqueueRequestForOwner{
-		IsController: true,
-		OwnerType:    &buildv1alpha1.BuildRun{},
-	})
+	// enqueue Reconciles requests only for events where a TaskRun already exists and that is related
+	// to a BuildRun
+	return c.Watch(&source.Kind{Type: &v1beta1.TaskRun{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(o handler.MapObject) []reconcile.Request {
+			taskRun := o.Object.(*v1beta1.TaskRun)
+
+			mgrClient := mgr.GetClient()
+			taskRunList := &v1beta1.TaskRunList{}
+
+			lbls := map[string]string{
+				buildv1alpha1.LabelBuildRun: taskRun.Name,
+			}
+			opts := client.ListOptions{
+				Namespace:     taskRun.Namespace,
+				LabelSelector: labels.SelectorFromSet(lbls),
+			}
+			_ = mgrClient.List(ctx, taskRunList, &opts)
+
+			res := make([]reconcile.Request, len(taskRunList.Items))
+			for i, tr := range taskRunList.Items {
+				res[i].Name = tr.Name
+				res[i].Namespace = tr.Namespace
+			}
+			return res
```

Comment:
Can't this be shortened to:

```go
			// cast object to TaskRun
			taskRun := o.Object.(*v1beta1.TaskRun)

			// check if TaskRun is related to BuildRun
			if taskRun.GetLabels() == nil || taskRun.GetLabels()[buildv1alpha1.LabelBuildRun] == "" {
				return make([]reconcile.Request, 0)
			} else {
				res := make([]reconcile.Request, 1)
				res[0].Name = taskRun.GetLabels()[buildv1alpha1.LabelBuildRun]
				res[0].Namespace = taskRun.Namespace

				return res
			}

```

Reply:
definitely, thanks for the suggestion. Done